src/Credentials/CredentialProvider.php

```diff
@@ -154,6 +156,36 @@ public static function memoize(callable $provider)
         };
     }
 
+    public static function cache(
+        callable $provider,
+        CacheInterface $cache = null,
+        $cacheKey = null
+    ) {
+        $cache = $cache ?: new FileCache;
+        $cacheKey = $cacheKey ?: 'aws_cached_credentials';
+
+        if (!$cache) {
+            return $provider;
+        }
+
+        return function () use ($provider, $cache, $cacheKey) {
+            $found = $cache->get($cacheKey);
+            if ($found instanceof CredentialsInterface && !$found->isExpired()) {
+                return Promise\promise_for($found);
+            }
+
+            $cacheable = $provider()->wait();
```

Comment:
This isn't what we want. We want to call `then` on the provider and return a promise that is fulfilled with the credentials. In that callback we'd then set the credentials in the cache.